control-plane/connect-inject/endpoints_controller.go

```diff
@@ -834,6 +835,168 @@ func serviceInstancesForK8SServiceNameAndNamespace(k8sServiceName, k8sServiceNam
+
+	// If Consul Namespaces or Admin Partitions are enabled, attempt to parse the
+	// upstream for a namespace.
+	if r.EnableConsulNamespaces || r.EnableConsulPartitions {
+		pieces := strings.SplitN(parts[0], ".", 3)
+		switch len(pieces) {
+		case 3:
+			partition = strings.TrimSpace(pieces[2])
+			fallthrough
+		case 2:
+			namespace = strings.TrimSpace(pieces[1])
+			fallthrough
+		default:
+			serviceName = strings.TrimSpace(pieces[0])
+		}
+	} else {
+		serviceName = strings.TrimSpace(parts[0])
+	}
+
+	// parse the optional datacenter
+	if len(parts) > 2 {
+		datacenter = strings.TrimSpace(parts[2])
+
+		// Check if there's a proxy defaults config with mesh gateway
+		// mode set to local or remote. This helps users from
+		// accidentally forgetting to set a mesh gateway mode
+		// and then being confused as to why their traffic isn't
+		// routing.
+		entry, _, err := r.ConsulClient.ConfigEntries().Get(api.ProxyDefaults, api.ProxyConfigGlobal, nil)
+		if err != nil && strings.Contains(err.Error(), "Unexpected response code: 404") {
+			return api.Upstream{}, fmt.Errorf("upstream %q is invalid: there is no ProxyDefaults config to set mesh gateway mode", rawUpstream)
+		} else if err == nil {
+			mode := entry.(*api.ProxyConfigEntry).MeshGateway.Mode
+			if mode != api.MeshGatewayModeLocal && mode != api.MeshGatewayModeRemote {
+				return api.Upstream{}, fmt.Errorf("upstream %q is invalid: ProxyDefaults mesh gateway mode is neither %q nor %q", rawUpstream, api.MeshGatewayModeLocal, api.MeshGatewayModeRemote)
+			}
+		}
+		// NOTE: If we can't reach Consul we don't error out because
+		// that would fail the pod scheduling and this is a nice-to-have
+		// check, not something that should block during a Consul hiccup.
+	}
+	if port > 0 {
+		upstream = api.Upstream{
+			DestinationType:      api.UpstreamDestTypeService,
+			DestinationPartition: partition,
+			DestinationPeer:      peer,
+			DestinationNamespace: namespace,
+			DestinationName:      serviceName,
+			Datacenter:           datacenter,
+			LocalBindPort:        int(port),
+		}
+	}
+	return upstream, nil
+
+}
+
+func (r *EndpointsController) processAnnotatedUpstream(pod corev1.Pod, rawUpstream string) (api.Upstream, error) {
+	var datacenter, serviceName, namespace, partition, peer string
+	var port int32
+	var upstream api.Upstream
+
+	parts := strings.SplitN(rawUpstream, ":", 3)
+
+	port, _ = portValue(pod, strings.TrimSpace(parts[1]))
+
+	service := parts[0]
+
+	pieces := strings.Split(service, ".")
+
+	if r.EnableConsulNamespaces || r.EnableConsulPartitions {
+		switch len(pieces) {
+		case 6:
+			end := strings.TrimSpace(pieces[5])
+			switch end {
+			case "peer":
+				peer = strings.TrimSpace(pieces[4])
+			case "ap":
+				partition = strings.TrimSpace(pieces[4])
+			case "dc":
+				datacenter = strings.TrimSpace(pieces[4])
+			default:
+				return api.Upstream{}, errors.New(fmt.Sprintf("upstream structured incorrectly: %s", rawUpstream))
```

Comment:
I wonder if this should be a part of the Validation logic that we do in the connect webhook? that way, if the label is incorrect we can ensure pod creation fails with an accurate error message.

Reply:
Yeah, I think in order to do that we'd need to have the entire logic for this function live in the webhook since you can't catch these cases without parsing the entire annotation. I think it'd work, and it'd probably be nice to have the pod creation fail, so do you think we should have the logic in both places? The prepared query case and the "non-annotated" upstreams are just less likely to have parsing issues compared to this one due to the number of pieces to parse.

Reply:
If the two are in the same package, we should be able to extract this behavior in a way that can be reused potentially. I also think it is something we can do down the line as a nice to have. Let's get this merged in and create an asana task for the validation if that works for you as well!

Reply:
Yup totally makes sense, added an Asana